service/worker/scanner/tasklist/handler.go

```diff
@@ -47,60 +48,84 @@ const scannerTaskListPrefix = "cadence-sys-tl-scanner"
 // with the assumption that the executor will schedule this task later
 //
 // Each loop of the handler proceeds as follows
-//    - Retrieve the next batch of tasks sorted by task_id for this task-list from persistence
+//    - Attempt to delete tasks up to the persisted ACK level.
 //    - If there are 0 tasks for this task-list, try deleting the task-list if its idle
-//    - If any of the tasks in the batch isn't expired, we are done. Since tasks are retrieved
-//      in sorted order, if one of the tasks isn't expired, chances are, none of the tasks above
-//      it are expired as well - so, we give up and wait for the next run
-//    - Delete the entire batch of tasks
-//    - If the number of tasks retrieved is less than batchSize, there are no more tasks in the task-list
+//    - If the number of tasks completed is less than the batchSize, there are no more tasks in the task-list
 //      Try deleting the task-list if its idle
-func (s *Scavenger) deleteHandler(key *taskListKey, state *taskListState) handlerStatus {
+func (s *Scavenger) deleteHandler(info *p.TaskListInfo) handlerStatus {
 	var err error
 	var nProcessed, nDeleted int
 
-	defer func() { s.deleteHandlerLog(key, state, nProcessed, nDeleted, err) }()
+	taskBatchSize := s.taskBatchSizeFn()
+	max := s.maxTasksPerJobFn()
 
-	for nProcessed < maxTasksPerJob {
-		resp, err1 := s.getTasks(key, taskBatchSize)
+	defer func() { s.deleteHandlerLog(info, nProcessed, nDeleted, err) }()
+
+	for nProcessed < max {
+		// First we complete tasks up to the persisted acklevel regardless of the task expiration
+		nTasks, err := s.completeTasks(info, info.AckLevel, taskBatchSize)
+		if err == p.ErrPersistenceLimitExceeded {
+			s.logger.Info("scavenger.deleteHandler query was ratelimited; will retry")
+			return handlerStatusDefer
+		}
+		if err != nil {
+			s.logger.Error(fmt.Sprintf("Scavenger error completing tasks: %v", err))
+			return handlerStatusErr
+		}
+		nProcessed += nTasks
+		nDeleted += nTasks
+		if nTasks < taskBatchSize {
+			break
+		}
+	}
+	for nProcessed < max {
+		// if we finished completing tasks below the ack level, but still have budget, then focus on expired tasks
+		resp, err1 := s.getTasks(info, taskBatchSize)
 		if err1 != nil {
 			err = err1
+			s.logger.Error(fmt.Sprintf("Scavenger error getting tasks: %v", err))
```

Comment:
We want to use tag.Error(err) instead of `Sprintf`. 
This is because having a stable/static message string will make it possible for us to use it to build some dashboard